File: android/app/src/main/kotlin/com/pulsetrack/app/DeepLinkChannel.kt
package com.pulsetrack.app

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

class DeepLinkChannel : EventChannel.StreamHandler {
    private var sink: EventChannel.EventSink? = null
    private var initialLink: String? = null

    fun register(messenger: BinaryMessenger) {
        EventChannel(messenger, "pulsetrack/deep_links").setStreamHandler(this)
        MethodChannel(messenger, "pulsetrack/deep_link_method").setMethodCallHandler { call, result ->
            if (call.method == "initialLink") result.success(initialLink) else result.notImplemented()
        }
    }

    fun receive(uri: String) {
        if (initialLink == null) initialLink = uri
        sink?.success(uri)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        initialLink?.let { events?.success(it) }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: ios/Runner/DeepLinkChannel.swift
import Flutter
import UIKit

final class DeepLinkChannel: NSObject, FlutterStreamHandler {
    static let shared = DeepLinkChannel()
    private var sink: FlutterEventSink?
    private var initialLink: String?

    static func register(with controller: FlutterViewController) {
        let events = FlutterEventChannel(name: "pulsetrack/deep_links", binaryMessenger: controller.binaryMessenger)
        events.setStreamHandler(shared)
        let methods = FlutterMethodChannel(name: "pulsetrack/deep_link_method", binaryMessenger: controller.binaryMessenger)
        methods.setMethodCallHandler { call, result in
            if call.method == "initialLink" {
                result(shared.initialLink)
            } else {
                result(FlutterMethodNotImplemented)
            }
        }
    }

    func receive(_ url: URL) {
        let value = url.absoluteString
        if initialLink == nil { initialLink = value }
        sink?(value)
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        if let initialLink { events(initialLink) }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
